File: packages/core/src/tiers.ts
/**
 * Price tiers — the "fluid" cost and margin model.
 *
 * ── Why this file exists ────────────────────────────────────────────────────────
 * An earlier version of this engine used ONE shipping assumption (insured Priority
 * Express, ~$32 floor) and ONE margin floor (12%) at every price point. Both were
 * wrong, and together they produced a fake conclusion: that nothing under ~$800 was
 * tradeable.
 *
 * Reality is fluid in two directions at once:
 *
 *   • COSTS scale with value. A $30 watch ships in a padded envelope for $7.50, not
 *     insured Express for $40. Applying luxury logistics to a cheap watch invents a
 *     loss that does not exist.
 *
 *   • REQUIRED MARGIN % FALLS as value rises, because what actually has to be
 *     covered is an absolute number — your handling time and your risk — not a
 *     percentage. 35% of a $40 watch is $14. 1% of a $10,000 watch is $100. The
 *     second is the better trade, and the percentage says the opposite.
 *
 * So each tier carries its own shipping, packaging, margin floor, MINIMUM ABSOLUTE
 * GROSS, handling time and permitted payment rails. A deal must clear both the
 * percentage and the dollar floor for its tier.
 *
 * ── The rail is what makes the top end work ─────────────────────────────────────
 * Card processing is 2.9% + $0.30 — uncapped. On a $10,000 order that is $290.30,
 * which is three times a 1% gross margin. Stripe ACH Direct Debit is 0.8% CAPPED AT
 * $5 (the cap binds above $625). Same order: $5.
 *
 * Thin high-value margins are not merely allowed by the rail, they are CREATED by
 * it. That is why the top tiers require ACH or wire rather than merely preferring
 * it. It also cuts chargeback exposure, which is the other reason 1% is survivable
 * up there. See docs/12.
 */

export type PaymentRail = 'CARD' | 'ACH' | 'WIRE';

export interface PriceTier {
  label: string;
  /** Inclusive lower bound on ORDER VALUE (what the customer pays). */
  minOrderUsd: number;
  /** Exclusive upper bound. */
  maxOrderUsd: number;
  outboundShippingUsd: number;
  packagingUsd: number;
  /** Percentage floor. Falls as value rises. */
  minMarginPct: number;
  /** Absolute floor. This is what really gates the cheap end. */
  minGrossProfitUsd: number;
  /** Realistic intake + photography + packing time. Drives $/hour. */
  handlingMinutes: number;
  /**
   * Whether the customer pays shipping on top of the list price.
   *
   * At the cheap end this is decisive: $7.50 of postage on a $27 watch is 28% of the
   * sale, and absorbing it makes the tier look unviable when it simply isn't. Buyers
   * expect to pay postage on a cheap item and expect free shipping on an expensive
   * one, so the flag follows the convention rather than fighting it.
   */
  shippingChargedToCustomer: boolean;
  allowedRails: readonly PaymentRail[];
  preferredRail: PaymentRail;
  notes: string;
}

/**
 * Defaults. Every number here is a starting guess to be re-fitted from your own
 * closed sales — that is what `comp_snapshots` and the rejection log are for.
 */
export const PRICE_TIERS: readonly PriceTier[] = [
  {
    label: 'MICRO',
    minOrderUsd: 0,
    maxOrderUsd: 150,
    outboundShippingUsd: 7.5,
    packagingUsd: 1.5,
    minMarginPct: 0.3,
    minGrossProfitUsd: 10,
    handlingMinutes: 8,
    shippingChargedToCustomer: true,
    allowedRails: ['CARD'],
    preferredRail: 'CARD',
    notes:
      'USPS Ground Advantage, tracking + $100 insurance included. No authentication ' +
      'available or needed — a dispute costs less than the shipping. Skip the full ' +
      'intake SOP: 5 photos, no video. Gated on $/hour, not margin.',
  },
  {
    label: 'BUDGET',
    minOrderUsd: 150,
    maxOrderUsd: 400,
    outboundShippingUsd: 11,
    packagingUsd: 3,
    minMarginPct: 0.22,
    minGrossProfitUsd: 35,
    handlingMinutes: 15,
    shippingChargedToCustomer: true,
    allowedRails: ['CARD'],
    preferredRail: 'CARD',
    notes: 'Ground Advantage with added insurance. Signature optional. 10 photos.',
  },
  {
    label: 'ENTRY',
    minOrderUsd: 400,
    maxOrderUsd: 1000,
    outboundShippingUsd: 22,
    packagingUsd: 6,
    minMarginPct: 0.15,
    minGrossProfitUsd: 90,
    handlingMinutes: 30,
    shippingChargedToCustomer: false,
    allowedRails: ['CARD'],
    preferredRail: 'CARD',
    notes:
      'Priority Mail, signature required, insured. Authenticity Guarantee add-on ' +
      'available once the SOURCE price clears $500. Full intake SOP starts here.',
  },
  {
    label: 'CORE',
    minOrderUsd: 1000,
    maxOrderUsd: 2500,
    outboundShippingUsd: 40,
    packagingUsd: 9,
    minMarginPct: 0.12,
    minGrossProfitUsd: 120,
    handlingMinutes: 35,
    shippingChargedToCustomer: false,
    allowedRails: ['CARD', 'ACH'],
    preferredRail: 'CARD',
    notes:
      'Priority Express, adult signature, insured. The bread-and-butter band: ' +
      'deep enough spreads to work, small enough that one bad unit is survivable.',
  },
  {
    label: 'UPPER',
    minOrderUsd: 2500,
    maxOrderUsd: 7500,
    outboundShippingUsd: 55,
    packagingUsd: 12,
    minMarginPct: 0.06,
    minGrossProfitUsd: 300,
    handlingMinutes: 45,
    shippingChargedToCustomer: false,
    allowedRails: ['CARD', 'ACH'],
    preferredRail: 'ACH',
    notes:
      'Authenticity Guarantee is free at $2,000+. Card fees start to bite hard ' +
      '(~$220 on a $7,500 order) — offer an ACH discount and most buyers take it.',
  },
  {
    label: 'HIGH',
    minOrderUsd: 7500,
    maxOrderUsd: 25000,
    outboundShippingUsd: 110,
    packagingUsd: 20,
    minMarginPct: 0.01,
    minGrossProfitUsd: 100,
    handlingMinutes: 60,
    shippingChargedToCustomer: false,
    allowedRails: ['ACH', 'WIRE'],
    preferredRail: 'WIRE',
    notes:
      'Registered Mail or Parcel Pro. CARD IS NOT PERMITTED: 2.9% uncapped would ' +
      'exceed the entire gross margin. 1% of $10,000 is $100 and that is a real ' +
      'profit — it only works because the rail is cheap and hard to reverse.',
  },
  {
    label: 'ULTRA',
    minOrderUsd: 25000,
    maxOrderUsd: Number.POSITIVE_INFINITY,
    outboundShippingUsd: 200,
    packagingUsd: 30,
    minMarginPct: 0.01,
    minGrossProfitUsd: 250,
    handlingMinutes: 90,
    shippingChargedToCustomer: false,
    allowedRails: ['WIRE'],
    preferredRail: 'WIRE',
    notes:
      'Wire only, funds cleared before the watch moves. Parcel Pro or armoured ' +
      'courier. Every unit is a bespoke transaction with a phone call.',
  },
];

export function tierFor(orderValueUsd: number): PriceTier {
  const found = PRICE_TIERS.find(
    (t) => orderValueUsd >= t.minOrderUsd && orderValueUsd < t.maxOrderUsd,
  );
  // Only reachable for negative input; treat as the cheapest tier.
  return found ?? PRICE_TIERS[0]!;
}

/** Processing cost for an order value on a given rail. */
export function processingFeeUsd(
  orderValueUsd: number,
  rail: PaymentRail,
  opts: {
    cardPercent?: number;
    cardFixedUsd?: number;
    radarUsd?: number;
    achPercent?: number;
    achCapUsd?: number;
    wireFeeUsd?: number;
  } = {},
): number {
  const {
    cardPercent = 0.029,
    cardFixedUsd = 0.3,
    radarUsd = 0.07,
    achPercent = 0.008,
    achCapUsd = 5,
    wireFeeUsd = 0,
  } = opts;

  switch (rail) {
    case 'ACH':
      // 0.8%, capped at $5 — the cap binds above $625.
      return Math.min(orderValueUsd * achPercent, achCapUsd);
    case 'WIRE':
      return wireFeeUsd;
    case 'CARD':
    default:
      return orderValueUsd * cardPercent + cardFixedUsd + radarUsd;
  }
}

/**
 * The rail we should actually use: the tier's preference, unless the customer's
 * choice is permitted at this value. Returns null when the requested rail is banned
 * — which is a hard stop, not a warning.
 */
export function resolveRail(
  orderValueUsd: number,
  requested?: PaymentRail,
): PaymentRail | null {
  const tier = tierFor(orderValueUsd);
  if (!requested) return tier.preferredRail;
  return tier.allowedRails.includes(requested) ? requested : null;
}

File: services/scout/scout/tiers.py
"""Price tiers — the fluid cost and margin model.

A port of packages/core/src/tiers.ts. See that file for the full reasoning; the short
version:

  • COSTS scale with value. A $30 watch ships in a $7.50 padded envelope, not insured
    Express for $40. Applying luxury logistics to a cheap watch invents a loss.

  • REQUIRED MARGIN % FALLS as value rises, because what must be covered is an
    absolute number — handling time and risk — not a percentage. 30% of a $34 order
    is $10. 1% of a $10,000 order is $100. The second is the better trade and a flat
    percentage floor says the opposite.

  • THE RAIL MAKES THE TOP END POSSIBLE. Card is 2.9% + $0.30 uncapped: $290 on a
    $10,000 order, nearly three times a 1% gross margin. Stripe ACH is 0.8% capped at
    $5. Thin high-value margins are created by the rail, which is why the top tiers
    require ACH or wire rather than merely preferring it.
"""

from __future__ import annotations

from dataclasses import dataclass, field

CARD = "CARD"
ACH = "ACH"
WIRE = "WIRE"


@dataclass(frozen=True)
class PriceTier:
    label: str
    min_order_usd: float
    max_order_usd: float
    outbound_shipping_usd: float
    packaging_usd: float
    min_margin_pct: float
    min_gross_profit_usd: float
    handling_minutes: int
    allowed_rails: tuple[str, ...]
    preferred_rail: str
    # Cheap tiers collect postage on top of the item price; expensive tiers absorb it.
    # $7.50 of postage on a $27 watch is 28% of the sale — absorbing it makes the tier
    # look unviable when it simply isn't.
    shipping_charged_to_customer: bool
    notes: str = ""


PRICE_TIERS: tuple[PriceTier, ...] = (
    PriceTier("MICRO", 0, 150, 7.50, 1.50, 0.30, 10, 8, (CARD,), CARD, True,
              "Ground Advantage, tracking + $100 insurance included. No authentication "
              "available or needed. 5 photos, no video. Gated on $/hour, not margin."),
    PriceTier("BUDGET", 150, 400, 11.0, 3.0, 0.22, 35, 15, (CARD,), CARD, True,
              "Ground Advantage with added insurance. Signature optional. 10 photos."),
    PriceTier("ENTRY", 400, 1000, 22.0, 6.0, 0.15, 90, 30, (CARD,), CARD, False,
              "Priority Mail, signature, insured. Authenticity Guarantee add-on once "
              "the SOURCE price clears $500. Full intake SOP starts here."),
    PriceTier("CORE", 1000, 2500, 40.0, 9.0, 0.12, 120, 35, (CARD, ACH), CARD, False,
              "Priority Express, adult signature, insured. The bread-and-butter band."),
    PriceTier("UPPER", 2500, 7500, 55.0, 12.0, 0.06, 300, 45, (CARD, ACH), ACH, False,
              "Authenticity Guarantee free at $2,000+. Card fees bite hard here — "
              "offer an ACH discount and most buyers take it."),
    PriceTier("HIGH", 7500, 25000, 110.0, 20.0, 0.01, 100, 60, (ACH, WIRE), WIRE, False,
              "Registered Mail or Parcel Pro. CARD NOT PERMITTED: 2.9% uncapped would "
              "exceed the entire gross margin. 1% of $10,000 is $100 and that is real "
              "profit — it works because the rail is cheap and hard to reverse."),
    PriceTier("ULTRA", 25000, float("inf"), 200.0, 30.0, 0.01, 250, 90, (WIRE,), WIRE, False,
              "Wire only, funds cleared before the watch moves. Every unit is bespoke."),
)


def tier_for(order_value_usd: float) -> PriceTier:
    for t in PRICE_TIERS:
        if t.min_order_usd <= order_value_usd < t.max_order_usd:
            return t
    return PRICE_TIERS[0]


def processing_fee(order_value_usd: float, rail: str, cfg) -> float:
    """Processing cost for an order value on a given rail."""
    if rail == ACH:
        # 0.8% capped at $5 — the cap binds above $625.
        return min(order_value_usd * cfg.ach_percent, cfg.ach_cap_usd)
    if rail == WIRE:
        return cfg.wire_fee_usd
    return order_value_usd * cfg.card_percent + cfg.card_fixed_usd + cfg.radar_per_txn_usd


def resolve_rail(order_value_usd: float, requested: str | None = None) -> str | None:
    """The rail to actually use. None means the requested rail is banned at this value."""
    tier = tier_for(order_value_usd)
    if requested is None:
        return tier.preferred_rail
    return requested if requested in tier.allowed_rails else None
